File: IOS/TMStore/TMStore/Classes/Models/CurrencyHelper.m
//
//  CurrencyHelper.m
//  TMStore
//
//  Created by Rishabh Jain on 27/12/16.
//  Copyright © 2016 Twist Mobile. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "CurrencyHelper.h"
#import "Utility.h"
#import "CurrencyItem.h"
#import "TM_CommonInfo.h"

static NSString* currencyName = @"";
static CurrencyItem *currencyItem;

static TM_CommonInfo *tm_commonInfo;

@implementation CurrencyHelper

- (id)init {
    self = [super init];
    if (self) {

    }
    return self;
}
+ (void)setSelectedCurrencyItem:(CurrencyItem*)currency_Item{
    currencyItem = currency_Item;
    currencyName = [currencyItem name];
}

+ (void)setCurrencyName:(NSString*)currency_Name{
    currencyName = currency_Name;
    for (CurrencyItem * cItem in [CurrencyItem getAll]) {
        if ([cItem.name isEqualToString:currencyName]) {
            currencyItem = cItem;
        }
    }
}

+ (CurrencyItem*)getCurrencyItemWithName:(NSString*)currency_Name{
    for (CurrencyItem *cItem in [CurrencyItem getAll]) {
        if ([cItem.name isEqualToString:currency_Name]) {
            return cItem;
        }
    }
    return nil;
}
+ (void)applyCurrencyRate:(ProductInfo*)product{
    if (currencyItem == nil) {
        return;
    }
    tm_commonInfo.currency_format = [currencyItem symbol];
    tm_commonInfo.currency = [currencyItem name];
    NSLog(@"Product price before currency change: %f",product._price);
    product._price = [self applyRate:product._price];
    product._regular_price = [self applyRate:product._regular_price];
    product._sale_price = [self applyRate:product._sale_price];
    NSLog(@"Product price after currency change: %f",product._price);

}
+ (void)applyCurrencyRateForVariation:(Variation*)variation{
    if (currencyItem == nil) {
        return;
    }
    tm_commonInfo.currency_format = [currencyItem symbol];
    tm_commonInfo.currency = [currencyItem name];
    NSLog(@"Variation price before currency change: %f",variation._price);
    variation._price = [self applyRate:variation._price];
    variation._regular_price = [self applyRate:variation._regular_price];
    variation._sale_price = [self applyRate:variation._sale_price];
    NSLog(@"Variation price after currency change: %f",variation._price);

}
+ (float)applyRate:(float)price{
    return currencyItem == nil ? price : price * [currencyItem rate];
}
+ (float)parseSafeFloatPrice:(NSString*)input{
    if (input != nil && [input floatValue]> 0) {
        @try {
            return [input floatValue];
        } @catch (NSException *e1) {
            @try {
                if ([input containsString:@","]) {
                    input = [input stringByReplacingOccurrencesOfString:@"," withString:@"."];
                }
                return [input floatValue];
            } @catch (NSException *e2) {
            }
        }
     }
    return 0.0f;
}
@end

File: Android/_setup/_ProjectName_/src/main/java/com/utils/CurrencyHelper.java
package com.utils;


import com.twist.dataengine.entities.CurrencyItem;
import com.twist.dataengine.entities.TM_CommonInfo;
import com.twist.dataengine.entities.TM_ProductInfo;
import com.twist.dataengine.entities.TM_Variation;
import com.twist.tmstore.entities.AppInfo;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Twist Mobile on 8/28/2017.
 */

public class CurrencyHelper {

    private static final String TAG = CurrencyHelper.class.getSimpleName();

    private static List<CurrencyItem> mCurrencyItems = new ArrayList<>();

    public static void clearAllCurrency() {
        if (mCurrencyItems != null) {
            mCurrencyItems.clear();
        }
    }

    public static void addNewCurrency(CurrencyItem currencyItem) {
        mCurrencyItems.add(currencyItem);
    }

    public static List<CurrencyItem> getAllCurrency() {
        return mCurrencyItems;
    }

    private static CurrencyItem currencyItem;
    private static String currencyName = "";

    public static void setSelectedCurrencyItem(CurrencyItem currencyItem) {
        CurrencyHelper.currencyItem = currencyItem;
        CurrencyHelper.currencyName = currencyItem.getName();
        TM_CommonInfo.currency_format = currencyItem.getSymbol();
        TM_CommonInfo.currency = currencyItem.getName();
    }

    public static void loadSavedCurrency() {
        if (!AppInfo.ENABLE_CURRENCY_SWITCHER)
            return;

        CurrencyHelper.currencyName = Preferences.getString(com.twist.tmstore.R.string.key_app_currency, TM_CommonInfo.currency);
        for (CurrencyItem currencyItem : getAllCurrency()) {
            if (currencyItem.getName().equalsIgnoreCase(CurrencyHelper.currencyName)) {
                CurrencyHelper.currencyItem = currencyItem;
                break;
            }
        }
    }

    public static CurrencyItem getCurrencyItemWithName(String currencyName) {
        for (CurrencyItem currencyItem : getAllCurrency()) {
            if (currencyItem.getName().equalsIgnoreCase(currencyName)) {
                return currencyItem;
            }
        }
        return null;
    }

    public static void applyCurrencyRate(TM_ProductInfo product) {
        if (currencyItem == null)
            return;

        TM_CommonInfo.currency_format = currencyItem.getSymbol();
        TM_CommonInfo.currency = currencyItem.getName();
        DataHelper.log(TAG + " Product Price before currency change => " + product.price);
        product.price = CurrencyHelper.applyRate(product.price);
        product.regular_price = CurrencyHelper.applyRate(product.regular_price);
        product.sale_price = CurrencyHelper.applyRate(product.sale_price);
        DataHelper.log(TAG + " Product Price after currency change => " + product.price);
    }

    public static void applyCurrencyRate(TM_Variation variation) {
        if (currencyItem == null)
            return;

        TM_CommonInfo.currency_format = currencyItem.getSymbol();
        TM_CommonInfo.currency = currencyItem.getName();
        DataHelper.log(TAG + " Variation Price before currency change => " + variation.price);
        variation.price = CurrencyHelper.applyRate(variation.price);
        variation.regular_price = CurrencyHelper.applyRate(variation.regular_price);
        variation.sale_price = CurrencyHelper.applyRate(variation.sale_price);
        DataHelper.log(TAG + " Variation Price after currency change => " + variation.price);
    }

    public static float applyRate(float price) {
        return currencyItem == null ? price : price * currencyItem.getRate();
    }
}
